### PR TITLE
json-schema-to-grammar improvements (+ added to server)

### PR DESCRIPTION
**Pull request by ochafik**
Improved JSON schema → GBNF grammar conversion support

- JSON schema features newly supported:
  - `required`, `allOf`, `anyOf`
  - [tuples](https://json-schema.org/understanding-json-schema/reference/array#tupleValidation) (incl. common legacy `items` syntax)
  - `$ref` (in same schema, or over https if `--allow-fetch` is set), allowing recursive types
  - `additionalProperties`
  - `pattern` ([most features](https://json-schema.org/understanding-json-schema/reference/regular_expressions), no greediness modifiers nor lookaheads; dot `.` can be made to match line breaks with `--dotall` flag)
  - `date`, `time`, `date-time`, `uuid` [string formats](https://json-schema.org/draft/2020-12/json-schema-validation#name-defined-formats)
- Ported to C++ (and updated JS version):
  - Added to `server`: the API now handles `"response_format": {"type": "json_object", "schema": ...}` on its own (see examples below), bridging an important gap between the C++ server and llama-cpp-python
  - Added tests to ensure the Python, JavaScript and C++ versions are synced. Their only differences are:
    - The C++ version has no option to resolve remote `$ref` (deemed too risky in server)
    - The Python version supports promoting a `{type: "string", pattern: "..."}` as a raw pattern (used by `examples/regex-to-grammar.py`, see example below)
- Fixes
  - Prevent `[, 1]`
  - Respect original property order (except for required properties guaranteed to appear before optionals, and `--prop-order` flag)

As a result, it can now consume the JSON schemas produced by [Pydantic](https://docs.pydantic.dev/) (used by ~ all Python LLM frameworks), [typescript-json-schema](https://github.com/YousefED/typescript-json-schema) and its more recent fork [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator), along with some more advanced schemas ([tsconfig.json](https://json.schemastore.org/tsconfig.json) is the toughest I tested).

Hopefully this PR hasn't grown too big, happy to send it in chunk if needed.

## Examples

(outputs below are with [Nous-Hermes-2-Mixtruct-v0.1-8x7B-DPO-DARE_TIES-Q6_K](https://huggingface.co/notstoic/Nous-Hermes-2-Mixtruct-v0.1-8x7B-DPO-DARE_TIES))

*   `tsconfig.json`

    ```bash
    ./main --grammar-file \
      <( python examples/json-schema-to-grammar.py https://json.schemastore.org/tsconfig.json ) \
      -p "Write a tsconfig.json for a simple project with strict types incremental compiler/build options:"
    ```
    <details>
      <summary>Show output</summary>

      ```js
      {
        "compilerOptions": {
          "module": "esnext",
          "target": "es2017",
          "sourceMap": true,
          "outDir": "./dist",
          "strict": true,
          "moduleResolution": "node",
          "esModuleInterop": true,
          "skipLibCheck": true,
          "forceConsistentCasingInFileNames": true,
          "lib": [
            "dom.iterable",
            "esnext"
          ],
          "declaration": false,
          "resolveJsonModule": true,
          "isolatedModules": true,
          "noEmit": false,
          "jsx": "react",
          "incremental": true
        },
        "include": [
          "./src/**/*.ts",
          "./src/**/*.tsx"
        ],
        "exclude": [
          "node_modules",
          "dist"
        ]
      }
      ```

    </details>

*   Pydantic w/ recursive types

    ```bash
    pip install pydantic
    
    echo "
    from pydantic import BaseModel
    from typing import Optional, Union, Tuple
    import json

    class QAPair(BaseModel):
      question: str
      concise_answer: str
      justification: str

    class PyramidalSummary(BaseModel):
      title: str
      summary: str
      question_answers: list[QAPair]
      sub_sections: list['PyramidalSummary']

    if __name__ == '__main__':
      print(json.dumps(PyramidalSummary.model_json_schema()))

    " | python - | tee qa-schema.json | python examples/json-schema-to-grammar.py - > qa-schema.gbnf

    ./main --grammar-file qa-schema.gbnf --log-disable --no-display-prompt -p "
      You are a highly efficient corporate document summarizer.
      Create a pyramidal summary of an imaginary internal document about our company processes
      (starting high-level, going down to each sub sections).
      Keep questions short, and answers even shorter (trivia / quizz style).
      Here is the schema of the output: $( cat qa-schema.json ).
    " | tee out.json
    ```

    <details>
      <summary>Show output</summary>

      ```js
      // cat out.json | jq
      {
        "title": "Company Processes Summary",
        "summary": "Our company has a strong focus on employee development, sustainability and innovation.",
        "question_answers": [
          {
            "question": "What is the main focus of our company?",
            "concise_answer": "Employee development, sustainability and innovation",
            "justification": "Mentioned in internal document."
          },
          {
            "question": "How does the company support employee development?",
            "concise_answer": "Regular training and mentorship programs.",
            "justification": "Document states 'We provide regular training and mentorship programs for employees to continually develop their skills.'"
          }
        ],
        "sub_sections": [
          {
            "title": "Employee Development Processes",
            "summary": "Our company provides regular training and mentorship programs for employees to grow professionally.",
            "question_answers": [
              {
                "question": "What methods does the company use to foster employee development?",
                "concise_answer": "Training workshops, online courses and mentorship.",
                "justification": "Stated in internal document."
              }
            ],
            "sub_sections": [
              {
                "title": "Training Workshops",
                "summary": "We organize regular training workshops to enhance employees' skills and knowledge.",
                "question_answers": [
                  {
                    "question": "What is the purpose of training workshops?",
                    "concise_answer": "To improve employees' skills and knowledge.",
                    "justification": "Said so in internal document."
                  }
                ],
                "sub_sections": []
              },
              {
                "title": "Online Courses",
                "summary": "We offer a variety of online courses for employees to learn at their own pace.",
                "question_answers": [
                  {
                    "question": "What is the advantage of offering online courses?",
                    "concise_answer": "Flexibility and convenience.",
                    "justification": "Can be inferred from document."
                  }
                ],
                "sub_sections": []
              },
              {
                "title": "Mentorship",
                "summary": "We pair employees with experienced mentors to guide them in their career paths.",
                "question_answers": [
                  {
                    "question": "Why do we have a mentorship program?",
                    "concise_answer": "To facilitate career development and growth.",
                    "justification": "Explicitly stated in document."
                  }
                ],
                "sub_sections": []
              }
            ]
          },
          {
            "title": "Sustainability Initiatives",
            "summary": "Our company is committed to promoting sustainability through various green initiatives.",
            "question_answers": [
              {
                "question": "What are some examples of our sustainability initiatives?",
                "concise_answer": "Recycling programs, energy-efficient lighting and eco-friendly office supplies.",
                "justification": "Mentioned in internal document."
              }
            ],
            "sub_sections": []
          },
          {
            "title": "Recycling Programs",
            "summary": "We have implemented recycling programs to minimize waste and promote sustainability.",
            "question_answers": [
              {
                "question": "What types of materials do we recycle?",
                "concise_answer": "Paper, plastic, glass and aluminum.",
                "justification": "Specified in internal document."
              }
            ],
            "sub_sections": []
          },
          {
            "title": "Energy-Efficient Lighting",
            "summary": "We use energy-efficient lighting solutions to reduce our carbon footprint and save on energy costs.",
            "question_answers": [
              {
                "question": "What kind of energy-efficient lighting do we use?",
                "concise_answer": "LED lights.",
                "justification": "Detailed in internal document."
              }
            ],
            "sub_sections": []
          },
          {
            "title": "Eco-Friendly Office Supplies",
            "summary": "We prioritize the use of eco-friendly office supplies to support our sustainability goals.",
            "question_answers": [
              {
                "question": "What are some examples of eco-friendly office supplies we use?",
                "concise_answer": "Recycled paper, non-toxic ink cartridges and reusable pens.",
                "justification": "Listed in internal document."
              }
            ],
            "sub_sections": []
          }
        ]
      }
      ```

    </details>

*   Regular expressions

    ```bash
    # Note that anything optional seems to require careful prompting.
    ./main --grammar "$( python examples/regex-to-grammar.py '^(\([0-9]{1,3}\))?[0-9]{3}-[0-9]{4}$' )" \
      -p "What is my phone number? (make sure to include the full prefix, wrapped in parentheses):"
    ```

    <details>
      <summary>Show output</summary>

      ```
      (970)512-4628
      ```

    </details>

*   TypeScript types

    ```bash
    # Notice optional 'e' accepts additional properties returning arrays of [string, number] tuples
    ./examples/ts-type-to-grammar.sh "{a: string, b: string, c?: number|string, d?: string, e?: {[i: string]: [string,number][]}}"
    ```

    <details>
      <summary>Show output</summary>

      ```js
      a-kv ::= "\"a\"" space ":" space string
      b-kv ::= "\"b\"" space ":" space string
      c ::= number | string
      c-kv ::= "\"c\"" space ":" space c
      c-rest ::= ( "," space d-kv )? d-rest
      d-kv ::= "\"d\"" space ":" space string
      d-rest ::= ( "," space e-kv )?
      e ::= "{" space  (e-additional-kvs )? "}" space
      e-additional-kv ::= string ":" space e-additional-value
      e-additional-kvs ::= e-additional-kv ( "," space e-additional-kv )*
      e-additional-value ::= "[" space ( e-additional-value-item ( "," space e-additional-value-item )* )? "]" space
      e-additional-value-item ::= "[" space string "," space number "]" space
      e-kv ::= "\"e\"" space ":" space e
      number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)? space
      root ::= "{" space a-kv "," space b-kv ( "," space ( c-kv c-rest | d-kv d-rest | e-kv ) )? "}" space
      space ::= " "?
      string ::=  "\"" (
              [^"\\] |
              "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F])
            )* "\"" space
      ```

    </details>

*   Typescript type including `Date`:

    ```bash
    # TypeScript allows for a compact description, useful to save tokens when putting the schema in the prompt itself:
    ( export TYPE="{name: string, description: string, start: Date, end?: Date}[]" ; \
 	    ./main --grammar-file <( ./examples/ts-type-to-grammar.sh "$TYPE" ) \
            --log-disable --no-display-prompt -n 2048 -p "
              List the 10 most noteworthy events or periods of the past 3 centuries, in JSON format '$TYPE'.
              Do specify an end date if the event was not an instant:
            " \
            | tee out.json )
    ```

    <details>
      <summary>Show output</summary>

      ```js
      // cat out.json | jq
      [
        {
          "name": "The Enlightenment",
          "description": "A cultural movement in Europe from roughly 1685 to 1815 that emphasized reason and individualism rather than tradition.",
          "start": "1685-01-01T00:00:00Z",
          "end": "1815-01-01T00:00:00Z"
        },
        {
          "name": "The Industrial Revolution",
          "description": "A period of major industrialization and innovation, starting in England around 1760 and lasting until the early 20th century.",
          "start": "1760-01-01T00:00:00Z",
          "end": "1920-01-01T00:00:00Z"
        },
        {
          "name": "The French Revolution",
          "description": "A period of radical social and political upheaval in France, starting in 1789 and lasting until 1799.",
          "start": "1789-01-01T00:00:00Z",
          "end": "1799-01-01T00:00:00Z"
        },
        {
          "name": "The American Civil War",
          "description": "A conflict between the Northern and Southern states of America, starting in 1861 and lasting until 1865.",
          "start": "1861-01-01T00:00:00Z",
          "end": "1865-01-01T00:00:00Z"
        },
        {
          "name": "The World Wars",
          "description": "Two major global conflicts in the 20th century, with World War I lasting from 1914 until 1918 and World War II from 1939 until 1945.",
          "start": "1914-01-01T00:00:00Z",
          "end": "1945-01-01T00:00:00Z"
        },
        {
          "name": "The Cold War",
          "description": "A period of geopolitical tension between the Western and Eastern blocs, starting in 1947 and lasting until 1991.",
          "start": "1947-01-01T00:00:00Z",
          "end": "1991-01-01T00:00:00Z"
        },
        {
          "name": "The Space Age",
          "description": "A period of space exploration and development, starting in the mid 20th century and continuing to this day.",
          "start": "1946-01-01T00:00:00Z",
          "end": "2100-01-01T00:00:00Z"
        }
      ]
      ```

    </details>

*   Other special formats:
    
    ```bash
    ./main \
      --grammar "$( \
        echo '{"prefixItems": [{ "format": "date" }, { "format": "uuid" }, { "format": "time" }, { "format": "date-time" }]}' | \
        python examples/json-schema-to-grammar.py - )" \
     -p "A very important date:"
    ```

    <details>
      <summary>Show output</summary>

      ```js
      [ "1967-05-29", "19680913-1968-0915-1968-091431140000", "19:00:00-19:01" ,"2008-07-30T11:16:29.154Z"] 
      ```

    </details>

*   JSON schema in server API (using [instructor](https://jxnl.github.io/instructor) **without** `llama-cpp-python`):

    ```bash
    pip install instructor openai pydantic
    ./server -m some-model.gguf &

    python -c '
    import instructor
    from typing import List
    from openai import OpenAI
    from pydantic import BaseModel
    
    client = instructor.patch(
        OpenAI(base_url="http://localhost:8080", api_key="123"),
        mode=instructor.Mode.JSON_SCHEMA)
    
    class UserDetail(BaseModel):
        name: str
        age: int
    
    print(client.chat.completions.create(
        model="whatever",
        messages=[{
            "role": "user",
            "content": "Extract `Jason is 30 years old`",
        }],
        response_model=UserDetail))
    '
    ```

    <details>
      <summary>Show output</summary>

      ```js
      name='Jason' age=30
      ```

    </details>

## TODOs (before undrafting this PR)

*   [x] Convert to `.mjs` again and test server + chat.mjs (had to implement a custom regexp parser instead of using Python's builtin one, to allow porting to JS)
*   [x] Test patterns w/ unicode and more syntactic features
*   [x] Document kv generation strategy (added longer example below)
*   [x] Support string format "date"
*   [x] Support string format "date-time"
*   [x] Test new OAI json format server behaviour
*   [x] Support mix of regular, optional and additional properties
*   ~~Fix sanitizer tests~~

## Possible followups:

*   [x] Support non-string `const` & `enum` → https://github.com/ggerganov/llama.cpp/pull/6232
*   [ ] Cap number length → https://github.com/ggerganov/llama.cpp/pull/6555
*   Support minimum, maximum, exclusiveMinimum, exclusiveMaximum at least for zero
*   Update / converge w/ `examples/pydantic-models-to-grammar-examples.py`
*   Publish `gbnf` Python package (similar to gguf) so `llama-cpp-python` needn't [copy / inline this script](https://github.com/abetlen/llama-cpp-python/blob/a7281994d87927e42d8e636295c786057e98d8fe/llama_cpp/llama_grammar.py#L1385)
*   Support string format "uri"

### CODE REVIEW

File: tests/test-json-schema-to-grammar.cpp [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  assert(std::system(("python ./examples/json-schema-to-grammar.py " + INPUT_NAME + " > " + OUT_NAME).c_str()) == 0);
+  tc.read_and_verify("Python");
+}
+
+static void run_mjs(const TestCase& tc) {
+  cerr << "# Running MJS " << tc.name.c_str() << endl;
+  tc.prepare();
+  assert(std::system(("node ./tests/run-json-schema-to-grammar.mjs " + INPUT_NAME + " > " + OUT_NAME).c_str()) == 0);
+  tc.read_and_verify("JavaScript");
+}
```
Comment (cebtenzzre) [inline tests/test-json-schema-to-grammar.cpp:72]: Doing this kind of string manipulation with the input to std::system is a really bad idea.

Comment (ochafik) [reply tests/test-json-schema-to-grammar.cpp]: Good point, looks smelly even in tests, now using `.tmp` files and static strings.

File: examples/server/json-schema-to-grammar.cpp [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+    {"string", " \"\\\"\" (\n"
+               "        [^\"\\\\] |\n"
+               "        \"\\\\\" ([\"\\\\/bfnrt] | \"u\" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F])\n"
+               "      )* \"\\\"\" space"},
+    {"null", "\"null\" space"}
+};
+vector<string> OBJECT_RULE_NAMES = {"object", "array", "string", "number", "boolean", "null", "value"};
+
+unordered_map<string, string> DATE_RULES = {
+    {"date", "[0-9] [0-9] [0-9] [0-9] \"-\" ( \"0\" [1-9] | \"1\" [0-2] ) \"-\" ( [0-2] [0-9] | \"3\" [0-1] )"},
```
Comment (ExtReMLapin) [inline examples/server/json-schema-to-grammar.cpp:39]: your date rule is incorrect, it can generate 00 for the day digits `[0-2] [0-9]`

Comment (ochafik) [reply examples/server/json-schema-to-grammar.cpp]: Oh, good catch, thanks! Fixed.

File: examples/server/json-schema-to-grammar.cpp [hunk 3]
```diff
@@ -0,0 +1,734 @@
...
+#include <map>
+#include <regex>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using json = nlohmann::json;
+using namespace std;
```
Comment (ggerganov) [inline examples/server/json-schema-to-grammar.cpp:14]: Avoid `using namespace std;`

`using json = ...` is fine

File: examples/server/json-schema-to-grammar.cpp [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+        result.append(replacement(match));
+        searchStart = match.suffix().first;
+    }
+
+    result.append(searchStart, searchEnd);
+
+    return result;
+}
+
+static string _format_literal(const string& literal) {
```
Comment (ggerganov) [inline examples/server/json-schema-to-grammar.cpp:128]: Global identifiers should not start with underscore

File: examples/server/json-schema-to-grammar.cpp [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+};
+
+string json_schema_to_grammar(const json& schema) {
+  SchemaConverter converter([](const string&) { return json::object(); }, /* dotall= */ false);
+  auto copy = schema;
+  converter.resolve_refs(copy, "input");
+  converter.visit(copy, "");
+  converter.check_errors();
+  return converter.format_grammar();
+}
```
Comment (ggerganov) [inline examples/server/json-schema-to-grammar.cpp:734]: Use 4-spaces indentation in C/C++ and Python sources

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -1,66 +1,75 @@
...
+llama_test(test-rope.cpp)
 
-llama_build_and_test_executable(test-rope.cpp)
+llama_test(test-model-load-cancel.cpp  LABEL "model")
+llama_test(test-autorelease.cpp        LABEL "model")
 
-llama_build_and_test_executable_with_label(test-model-load-cancel.cpp "model")
-llama_build_and_test_executable_with_label(test-autorelease.cpp "model")
+llama_test(test-json-schema-to-grammar.cpp   WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/..)
+target_include_directories(test-json-schema-to-grammar PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../examples/server)
```
Comment (ggerganov) [inline tests/CMakeLists.txt:72]: Move `json.hpp` and `json-schema-to-grammar.h` from `examples/server` to `common`

Comment (ochafik) [reply tests/CMakeLists.txt:72]: Done

File: examples/server/json-schema-to-grammar.cpp (hunk 5 above)
Comment (ochafik) [reply examples/server/json-schema-to-grammar.cpp]: Done

File: examples/server/json-schema-to-grammar.cpp (hunk 4 above)
Comment (ochafik) [reply examples/server/json-schema-to-grammar.cpp]: Fixed

File: examples/server/json-schema-to-grammar.cpp (hunk 3 above)
Comment (ochafik) [reply examples/server/json-schema-to-grammar.cpp]: std::done

File: tests/test-json-schema-to-grammar.cpp [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+    TestCaseStatus expected_status;
+    std::string name;
+    std::string schema;
+    std::string expected_grammar;
+
+    void _print_failure_header() const {
+        std::cerr << "#" << std::endl;
+        std::cerr << "# Test '" << name.c_str() << "' failed." << std::endl;
+        std::cerr << "#" << std::endl;
+        std::cerr << schema.c_str() << std::endl;
```
Comment (ggerganov) [inline tests/test-json-schema-to-grammar.cpp:35]: For consistency with the rest of the codebase:

`std::cerr -> fprintf(stderr,`
`std::cout -> fprintf(stdout,`

File: common/json-schema-to-grammar.cpp [hunk 8]
```diff
@@ -0,0 +1,726 @@
...
+    result.reserve(str.length() * n);
+
+    for (size_t i = 0; i < n; ++i) {
+        result += str;
+    }
+
+    return result;
+}
+
+static std::string replacePattern(const std::string& input, const std::regex& regex, const std::function<std::string(const std::smatch &)>& replacement) {
```
Comment (ggerganov) [inline common/json-schema-to-grammar.cpp:109]: For consistency, put spaces before and after - see the rest of the usages in the codebase:

```suggestion
static std::string replacePattern(const std::string & input, const std::regex & regex, const std::function<std::string(const std::smatch &)> & replacement) {
```

**Review by ggerganov — APPROVED** [review]
Squash merge when you are ready

I haven't tested the implementation, but it seems you've put enough effort and there are unit tests, so it should be good 👍

File: tests/test-json-schema-to-grammar.cpp (hunk 7 above)
Comment (ochafik) [reply tests/test-json-schema-to-grammar.cpp]: Done

File: common/json-schema-to-grammar.cpp (hunk 8 above)
Comment (ochafik) [reply common/json-schema-to-grammar.cpp]: Done

**Pull request merged**